Sprint 11

Skoch33 opened the pull request.

Skoch33 committed:
1. Добавил зависимость ProgressHUD
2. Устранил гонку в проекте
3. Добавили новый сервис ProfileService, для отображения актуальных данных пользователя

Skoch33 committed:
1. Поправил код
2. Добавил библиотеку Kingfisher с помощью SPM
2. Использовал Kingfisher для загрузки аватарки профиля из сети

Skoch33 committed:
no message

Skoch33 committed:
1. Фикс Баг

Skoch33 committed:
Заменил использование OAuth2TokenStorage на хранение и получение из хранилища токена с использованием SwiftKeychainWrapper

Skoch33 committed:
1. Создал ViewController из кода
2. Сверстал SplashViewController из кода

Helvetii951 commented on ImageFeed/Auth/OAuth2Service.swift:
```diff
@@ -8,86 +8,59 @@
+        currentTask?.cancel()
+        lastCode = code
+        
+        guard let request = makeRequest(code: code) else { return }
+        
+        let session = URLSession.shared
+        currentTask = session.objectTask(for: request) { [weak self] (result: Result<OAuthTokenResponseBody, Error>) in
+            guard let self else { return }
```
Желательно вызывать `completion` с ошибкой в `guard`ах

Helvetii951 commented on ImageFeed/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,66 @@
+import UIKit
+
+final class ProfileImageService {
+    static let shared = ProfileImageService()
```
Предполагается, что `ProfileImageService ` будет использоваться в одном экземпляре, т.е. через `shared`. Тогда стоит добавить приватный `init`. Так мы реализуем паттерн [singleton](https://www.hackingwithswift.com/example-code/language/what-is-a-singleton).

Helvetii951 commented on ImageFeed/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,67 @@
+import Foundation
+final class ProfileService {
+    
+    static let shared = ProfileService()
```
Тут тоже лучше сделать singleton. Я написал об этом в посте выше.

Helvetii951 commented on ImageFeed/Profile/ProfileModels.swift:
```diff
@@ -0,0 +1,15 @@
+
+import UIKit
+
+struct ProfileResult: Decodable {
+    let username: String
+    let first_name: String
+    let last_name: String
+    let bio: String?
+}
```
Для того, чтобы не работать с snake_case ключами, не описывать их в CodingKeys, ты можешь также 
настраивать JSONDecoder под тот вид данных, который приходит от сервера. 
Рекомендую изучить этот способ самостоятельно https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable

Helvetii951 commented on ImageFeed/Profile/ProfileModels.swift:
```diff
@@ -0,0 +1,15 @@
+struct ProfileResult: Decodable {
+    let username: String
+    let first_name: String
+    let last_name: String
```
`last_name` в `ProfileResult ` может и не быть в профиле. Свойство нужно сделать опциональным.

Helvetii951 commented on ImageFeed/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,67 @@
+        case unableToDecodeStringFromProfileData
+    }
+    
+    struct Profile {
+        let username: String
+        let name: String
+        let loginName: String
+        let bio: String
+    }
```
Эту структуру лучше вынести в отдельный файл для удобства навигации по проекту.  + она еще получилась [вложенным типом,](https://swiftbook.ru/content/languageguide/nested-types/) что не очень корректно.

Helvetii951 commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -52,48 +86,46 @@ final class SplashViewController: UIViewController {
+// MARK: - Alert
+
+extension SplashViewController {
+    func showAlert() {
+        let alert = UIAlertController(title: "Что-то пошло не так", message: "Не удалось войти в систему", preferredStyle: .alert)
+        let uiAlertAction = UIAlertAction(title: "Ok", style: .default) { _ in
+            alert.dismiss(animated: true)
+        }
+        alert.addAction(uiAlertAction)
+        present(alert, animated: true)
+    }
```
Функцию можно сделать `private`. Рекомендую писать функции и переменные с минимальной областью видимости, а потом менять, если необходимо.

Skoch33 committed:
1. Реализовал паттерн sigleton в ProfileImageService и ProfileService
2. Заменил Decodable на Codable в OAuthTokenResponseBody, ProfileResult и ProfileImageService
3. Сделал опциональным свойства first_name и last_name  в ProfileModels
4. Вынес структуру Profile в отдельный файл для удобства навигации по проекту
5. Сделал приватной функцию alert
